jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java

```diff
@@ -32,30 +40,83 @@ public class UdpServicesController {
-    serviceManager.submit(
-        new MapReceiver(udpProps.getMap(), kafkaTemplate, rawEncodedJsonTopics.getMap()));
-    serviceManager.submit(
-        new PsmReceiver(udpProps.getPsm(), kafkaTemplate, rawEncodedJsonTopics.getPsm()));
-    serviceManager.submit(
-        new GenericReceiver(udpProps.getGeneric(), kafkaTemplate, rawEncodedJsonTopics));
+    startReceiver(new BsmReceiver(udpProps.getBsm(), kafkaTemplate, rawEncodedJsonTopics.getBsm()));
+    startReceiver(new TimReceiver(udpProps.getTim(), kafkaTemplate, rawEncodedJsonTopics.getTim()));
+    startReceiver(new SsmReceiver(udpProps.getSsm(), kafkaTemplate, rawEncodedJsonTopics.getSsm()));
+    startReceiver(new SrmReceiver(udpProps.getSrm(), kafkaTemplate, rawEncodedJsonTopics.getSrm()));
+    startReceiver(new SpatReceiver(udpProps.getSpat(), kafkaTemplate, rawEncodedJsonTopics.getSpat()));
+    startReceiver(new MapReceiver(udpProps.getMap(), kafkaTemplate, rawEncodedJsonTopics.getMap()));
+    startReceiver(new PsmReceiver(udpProps.getPsm(), kafkaTemplate, rawEncodedJsonTopics.getPsm()));
+    startReceiver(new GenericReceiver(udpProps.getGeneric(), kafkaTemplate, rawEncodedJsonTopics));
 
     log.debug("UDP receiver services started.");
   }
-}
+
+  /**
+   * Starts a receiver in its own executor service and manages its lifecycle.
+   *
+   * @param receiver The receiver to start
+   */
+  private void startReceiver(Runnable receiver) {
+    ExecutorService executor = Executors.newSingleThreadExecutor(r -> {
+      Thread thread = new Thread(r);
+      thread.setDaemon(true); // Makes thread exit when main application exits
+      return thread;
+    });
+
+    executors.add(executor);
+    if (receiver instanceof AutoCloseable closeable) {
+      receivers.add(closeable);
+    }
+
+    executor.submit(() -> {
+      try {
+        while (!Thread.currentThread().isInterrupted()) {
+          receiver.run();
+        }
+      } catch (Exception e) {
+        log.error("Error in receiver {}: {}", receiver.getClass().getSimpleName(), e.getMessage(), e);
+      }
+    });
+  }
+
+  /**
+   * Gracefully shuts down all executor services and closes all receivers.
+   */
+  @PreDestroy
+  public void shutdown() {
+    log.info("Shutting down UDP services...");
+
+    // First, shutdown all executors
+    executors.forEach(executor -> {
+      try {
+        executor.shutdown();
+        if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
+          executor.shutdownNow();
+          if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
+            log.error("Executor did not terminate");
+          }
+        }
+      } catch (InterruptedException e) {
+        executor.shutdownNow();
+        Thread.currentThread().interrupt();
+      }
+    });
+
+    // Then, close all receivers
+    receivers.forEach(receiver -> {
+      try {
+        receiver.close();
+      } catch (Exception e) {
+        log.error("Error closing receiver: {}", e.getMessage(), e);
+      }
+    });
```

Comment:
This code never runs because there are no receivers in the list.

The base type AbstractUdpReceiverPublisher doesn't have a close method, but it does have a setStopped(boolean) method that could be called here.

Reply:
Resolved.  (I don't have permission to "resolve" comments.)